pkg/shared/callback/callback.go

```diff
@@ -0,0 +1,275 @@
+package callback
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"go.uber.org/zap"
+
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/isb"
+)
+
+// Publisher is the callback publisher which publishes the callback messages.
+type Publisher struct {
+	vertexName   string
+	pipelineName string
+	clientsCache *lru.Cache[string, *http.Client]
+	opts         *Options
+}
+
+// NewPublisher creates a new callback publisher
+func NewPublisher(ctx context.Context, vertexName string, pipelineName string, opts ...OptionFunc) *Publisher {
+	dOpts := DefaultOptions(ctx)
+	for _, opt := range opts {
+		opt(dOpts)
+	}
+
+	clientCache, _ := lru.NewWithEvict[string, *http.Client](dOpts.cacheSize, func(key string, value *http.Client) {
+		// Close the client when it's removed from the cache
+		value.CloseIdleConnections()
+	})
+
+	if dOpts.callbackURL != "" {
+		client := &http.Client{
```

Comment:
This has been existing in `GetClient()`, do not need it here?

Reply:
we don't need it